pkg/sql/plan2/function/builtin/unary/year.go

```diff
@@ -25,78 +25,103 @@ import (
 			return proc.AllocScalarNullVector(resultType), nil
 		}
 		resultVector := vector.NewConst(resultType)
-		resultValues := make([]uint16, 1)
-		vector.SetCol(resultVector, year.DateToYear(inputValues, resultValues))
+		resultValues := make([]int64, 1)
+		year.DateToYearPlan2(inputValues, resultValues)
+		// resultValues2 := make([]int64, 1)
+		// resultValues2[0] = int64(resultValues[0])
+		vector.SetCol(resultVector, resultValues)
 		return resultVector, nil
 	} else {
 		resultVector, err := proc.AllocVector(resultType, int64(resultElementSize*len(inputValues)))
 		if err != nil {
 			return nil, err
 		}
-		resultValues := encoding.DecodeUint16Slice(resultVector.Data)
+		resultValues := encoding.DecodeInt64Slice(resultVector.Data)
 		resultValues = resultValues[:len(inputValues)]
 		nulls.Set(resultVector.Nsp, inputVector.Nsp)
-		vector.SetCol(resultVector, year.DateToYear(inputValues, resultValues))
+		year.DateToYearPlan2(inputValues, resultValues)
+		// resultValues2 := make([]int64, len(resultValues))
+		// for i, x := range resultValues {
+		// 	resultValues2[i] = int64(x)
+		// }
+		vector.SetCol(resultVector, resultValues)
 		return resultVector, nil
 	}
 }
 
 func DatetimeToYear(vectors []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
 	inputVector := vectors[0]
-	resultType := types.Type{Oid: types.T_uint16, Size: 2}
+	resultType := types.Type{Oid: types.T_int64, Size: 8}
 	resultElementSize := int(resultType.Size)
 	inputValues := vector.MustTCols[types.Datetime](inputVector)
 	if inputVector.IsScalar() {
 		if inputVector.ConstVectorIsNull() {
 			return proc.AllocScalarNullVector(resultType), nil
 		}
 		resultVector := vector.NewConst(resultType)
-		resultValues := make([]uint16, 1)
-		vector.SetCol(resultVector, year.DatetimeToYear(inputValues, resultValues))
+		resultValues := make([]int64, 1)
+		year.DatetimeToYearPlan2(inputValues, resultValues)
+		// resultValues2 := make([]int64, 1)
+		// resultValues2[0] = int64(resultValues[0])
+		vector.SetCol(resultVector, resultValues)
 		return resultVector, nil
 	} else {
 		resultVector, err := proc.AllocVector(resultType, int64(resultElementSize*len(inputValues)))
 		if err != nil {
 			return nil, err
 		}
-		resultValues := encoding.DecodeUint16Slice(resultVector.Data)
+		// resultValues := make([]uint16, len(inputValues))
+		resultValues := encoding.DecodeInt64Slice(resultVector.Data)
 		resultValues = resultValues[:len(inputValues)]
 		nulls.Set(resultVector.Nsp, inputVector.Nsp)
-		vector.SetCol(resultVector, year.DatetimeToYear(inputValues, resultValues))
+		year.DatetimeToYearPlan2(inputValues, resultValues)
+		// resultValues2 := make([]int64, len(resultValues))
+		// for i, x := range resultValues {
+		// 	resultValues2[i] = int64(x)
+		// }
+		vector.SetCol(resultVector, resultValues)
 		return resultVector, nil
 	}
 }
 
 func DateStringToYear(vectors []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
```

Comment:
In general, this should be done by implicit casting and we don't need this function

Reply:
by casting varchar to date, but I guess it is okay to write this function explicitly